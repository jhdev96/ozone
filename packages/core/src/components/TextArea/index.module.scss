@import '../../scss/root';


$resizeList: none, both, horizontal, vertical;

.textarea {
  display: block;
  width: 100%;
  max-width: 100%;
  box-sizing: border-box;
  padding: 7px;
  border-radius: 4px;
  border: 1px solid var(--gray-4);
  font-family: var(--font-family);

  &:focus {
    border: 1px solid transparent !important;
    outline: none !important;
    box-shadow: 0 0 0 2px var(--blue-core);
    transition: 0.3s box-shadow;
  }
}

// theme
.light {
  color: var(--black-core);
  background-color: #fff;
}

.dark {
  color: var(--gray-1);
  background-color: var(--black-3);
}

.error { 
  color: var(--red-core);
  border: 1px solid transparent !important;
  box-shadow: 0 0 0 2px var(--red-core) !important;
}

// resize prop
@each $resizeType in $resizeList {
  .#{$resizeType} {
    resize: $resizeType;
  }
}