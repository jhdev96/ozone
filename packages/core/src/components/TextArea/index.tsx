import React, { useState } from 'react';
import BaseTextProps from '../../utils/BaseTextProps';
import classNames from '../../utils/classNames';
import { ThemeType } from '../../utils/theme';
import styles from './index.module.scss';


export interface TextAreaProps extends BaseTextProps<HTMLTextAreaElement> {
  /**
   * Applies additional css styling to the
   * textarea element.
   */
  className?: string;
  /**
   * Gives the component a light or dark theme.
   */
  theme?: ThemeType;
  /**
   * Sets the HTML `rows` attribute on the textarea
   * element.
   */
  minimumRows?: number;
  /**
   * Sets the HTML `defaultValue` attribute on the textarea
   * element.
   */
  defaultValue?: string;
  /**
   * Applies resize css styling to the textarea element.
   */
  resize?: "none" | "both" | "horizontal" | "vertical" | "auto";
};

const TextArea = React.forwardRef<HTMLTextAreaElement, TextAreaProps>(
  function TextArea({
    id,
    className,
    name,
    theme       = "light",
    isDisabled,
    isRequired,
    placeholder,
    hasError    = false,
    onChange    = () => {},
    onFocus     = () => {},
    onBlur      = () => {},
    onKeyDown   = () => {},
    onKeyPress  = () => {},
    onKeyUp     = () => {},
    resize      = "auto",
    minimumRows = 3,
    defaultValue}: TextAreaProps, 
    ref: React.ForwardedRef<HTMLTextAreaElement>): JSX.Element
  {
    const [value, setValue] = useState(defaultValue);

    const handleChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
      setValue(e.target.value);
      onChange(e, e.target.value);
    }

    return (
      <textarea 
        id={id}
        className={classNames(
          className,
          styles.textarea,
          styles[theme],
          styles[resize],
          hasError && styles.error
        )}
        name={name} 
        disabled={isDisabled}
        required={isRequired}
        ref={ref}
        value={value}
        placeholder={placeholder}
        rows={minimumRows}
        onChange={handleChange}
        onFocus={onFocus}
        onBlur={onBlur}
        onKeyDown={onKeyDown}
        onKeyPress={onKeyPress}
        onKeyUp={onKeyUp}
      />
    )
  }
)

TextArea.displayName = 'TextArea';

export default TextArea;