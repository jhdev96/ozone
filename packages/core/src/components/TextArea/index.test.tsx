import React from "react";
import { 
  render, 
  fireEvent 
} from "@testing-library/react";
import TextArea, { TextAreaProps } from ".";


describe('<TextArea />', () => {
  function setUp(props?: TextAreaProps, id="test"): HTMLTextAreaElement | null
  { 
    render(<TextArea id={id} {...props} />);

    return document.querySelector(`#${id}`);
  }

  it('renders a textarea with a placeholder', () => {
    const element = setUp({placeholder: "hello"});

    expect(element!.placeholder).toBe("hello");
  });

  it('renders a required textarea', () => {
    const element = setUp({isRequired: true});

    expect(element!.required).toBe(true);
  });

  it('renders a disabled textarea', () => {
    const element = setUp({isDisabled: true});

    expect(element!.disabled).toBe(true);
  });

  it('renders a textarea with a `name` attribute', () => {
    const element = setUp({name: 'TA'});

    expect(element!.name).toBe('TA');
  });

  it('renders a textarea with 4 rows', () => {
    const element = setUp({minimumRows: 4});

    expect(element!.rows).toBe(4);
  });

  it('sets the textarea\'s value to the `defaultValue` prop', () => {
    const element = setUp({defaultValue: 'hello'});

    expect(element!.value).toBe('hello');
  });

  it('sets the `resize` prop', () => {
    const element = setUp({resize: 'both'});

    expect(element).toHaveClass('both');
  });

  it('changes the textarea\'s value during change events', () => {
    const element = setUp();

    fireEvent.change(element!, {target: {value: 'hello'}});

    expect(element!.value).toBe('hello');
  });

  it('calls `onChange` during change events', () => {
    const onChange = jest.fn();
    const element = setUp({onChange: onChange});

    fireEvent.change(element!, {target: {value: 'hello'}});

    expect(onChange).toBeCalled();
  });

  it('calls `onFocus` when the textarea is focused', () => {
    const onFocus = jest.fn();
    const element = setUp({onFocus: onFocus});

    fireEvent.focus(element!);

    expect(onFocus).toBeCalled();
  });

  it('calls `onBlur` when the textarea loses focus', () => {
    const onBlur = jest.fn();
    const element = setUp({onBlur: onBlur});

    fireEvent.blur(element!);

    expect(onBlur).toBeCalled();
  });

  it('renders a textarea with an error class', () => {
    const element = setUp({hasError: true});

    expect(element).toHaveClass('error');
  });

  it('renders the correct css classes', () => {
    const element = setUp({resize: 'both'});

    expect(element).toMatchSnapshot();
  });
});