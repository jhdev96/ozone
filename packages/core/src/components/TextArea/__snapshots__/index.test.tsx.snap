// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextArea /> renders the correct css classes 1`] = `
<textarea
  class="textarea light both"
  id="test"
  rows="3"
/>
`;
